paddle/fluid/operators/dense_triplet_loss_op.h

```diff
@@ -0,0 +1,183 @@
+#include "paddle/fluid/operators/math/blas.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenVector = framework::EigenVector<T, MajorType, IndexType>;
+
+using DIM1 = Eigen::array<int, 1>;
+using DIM2 = Eigen::array<int, 2>;
+
+template <typename DeviceContext>
+std::vector<int> GetOffsets(const Tensor* t);
+
+template <typename T>
+class ReluFunctor {
+ public:
+  explicit ReluFunctor(T margin) : margin_(margin) {}
+  HOSTDEVICE T operator()(const T& x) const {
+    if ((x + margin_) > 0)
+      return x + margin_;
+    else
+      return 0;
+  }
+
+ private:
+  T margin_;
+};
+
+template <typename T>
+class ReluGradFunctor {
+ public:
+  HOSTDEVICE T operator()(const T& x) const { return x < 0 ? T(0) : T(1); }
+};
+
+template <typename DeviceContext, typename T>
+class DenseTripletLossKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto& place =
+        *context.template device_context<DeviceContext>().eigen_device();
+    const Tensor* logits = context.Input<Tensor>("Logits");
+    const Tensor* labels = context.Input<Tensor>("Label");
+    T margin = static_cast<T>(context.Attr<float>("margin"));
+    Tensor* loss = context.Output<Tensor>("Loss");
+    Tensor* d_logits = context.Output<Tensor>("LogitsGrad");
+    loss->mutable_data<T>(context.GetPlace());
+    d_logits->mutable_data<T>(context.GetPlace());
+    logits->data<T>();
+    auto x_dims = logits->dims();
+    int batch_size = x_dims[0];
+    int feature_len = x_dims[1];
+
+    // step 1: get distance matrix
+    Tensor distances;
+    distances.mutable_data<T>({batch_size, batch_size}, context.GetPlace());
+    Tensor d_distances;
+    d_distances.mutable_data<T>({batch_size, batch_size}, context.GetPlace());
+    Tensor tmp;
+    tmp.mutable_data<T>({batch_size, batch_size}, context.GetPlace());
+    math::SetConstant<DeviceContext, T> set_zero;
+    auto& dev_ctx = context.template device_context<DeviceContext>();
+    set_zero(dev_ctx, &d_distances, static_cast<T>(0));
+
+    auto logits_t = EigenMatrix<T>::From(*logits);
+    auto d_logits_t = EigenMatrix<T>::From(*d_logits);
+    auto loss_t = EigenMatrix<T>::From(*loss);
+    auto distances_t = EigenMatrix<T>::From(distances);
+    auto d_distances_t = EigenMatrix<T>::From(d_distances);
+    auto tmp_t = EigenMatrix<T>::From(tmp);
+
+    auto blas = math::GetBlas<DeviceContext, T>(context);
+    auto x_mat = math::CreateMatrixDescriptor(x_dims, 0, false);
+    auto x_mat_trans = math::CreateMatrixDescriptor(x_dims, 0, true);
+    blas.MatMul(*logits, x_mat, *logits, x_mat_trans, T(1), &distances, T(0));
```

Comment:
Since the input is 2D tensor, there is no need to use `math::CreateMatrixDescriptor`,  refer the MatMul inferface:
 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/math/blas_impl.h#L176